crates/noirc_evaluator/src/ssa/acir_gen/internal_var_cache.rs

```diff
@@ -76,10 +98,185 @@ impl InternalVarCache {
+    fn const_to_witness_helper(var: &mut InternalVar, evaluator: &mut Evaluator) -> Witness {
+        let w = Self::internal_get_or_compute_witness(var, evaluator);
+        if w.is_none() {
+            let witness = expression_to_witness(var.to_expression(), evaluator);
+            var.set_witness(Some(witness));
+        }
+        var.cached_witness().unwrap()
+    }
+
+    /// Get or compute a witness for an internal var
+    /// WARNING: It generates a witness even if the internal var is constant, so it should be used only if the var is an input
+    /// to some ACIR opcode which requires a witness
+    pub(crate) fn get_or_compute_witness_unwrap(
+        &mut self,
+        mut var: InternalVar,
+        evaluator: &mut Evaluator,
+        ctx: &SsaContext,
+    ) -> Witness {
+        if let Some(v) = var.to_const() {
+            self.const_to_witness(v, evaluator, ctx)
+        } else {
+            let w = Self::internal_get_or_compute_witness(&mut var, evaluator)
+                .expect("infaillible non const expression");
+            var.set_witness(Some(w));
+            self.update(var);
+            w
+        }
+    }
+
+    /// Get or compute a witness equating the internal var
+    /// It returns None when the variable is a constant instead of creating a witness
+    /// because we should not need a witness in that case
+    /// If you really need one, you can use get_or_compute_witness_unwrap()
+    pub(crate) fn get_or_compute_witness(
+        &mut self,
+        mut var: InternalVar,
+        evaluator: &mut Evaluator,
+    ) -> Option<Witness> {
+        let w = Self::internal_get_or_compute_witness(&mut var, evaluator);
+        if w.is_some() {
+            assert!(var.cached_witness().is_some());
+        } else {
+            return None;
+        };
+        self.update(var);
+
+        w
+    }
+
+    /// Generates a `Witness` that is equal to the `expression`.
+    /// - If a `Witness` has previously been generated, we return it.
+    /// - If the Expression represents a constant, we return None.
+    fn internal_get_or_compute_witness(
+        var: &mut InternalVar,
+        evaluator: &mut Evaluator,
+    ) -> Option<Witness> {
+        // Check if we've already generated a `Witness` which is equal to
+        // the stored `Expression`
+        if let Some(witness) = var.cached_witness() {
+            return Some(*witness);
+        }
+
+        // We do not generate a witness for constant values. It can only be done at the InternalVarCache level.
+        if var.is_const_expression() {
+            return None;
+        }
+
+        var.set_witness(Some(expression_to_witness(var.expression().clone(), evaluator)));
+
+        *var.cached_witness()
+    }
+
+    pub(super) fn update(&mut self, var: InternalVar) {
+        if let Some(id) = var.get_id() {
+            self.inner.insert(id, var);
+        } else if let Some(value) = var.to_const() {
+            self.constants.insert(value, var);
+        }
```

Comment:
If either of these branches do not execute, wouldn't that be a compiler error?

Reply:
No, it can happen when we generate expressions from internal vars